includes/breadcrumbs.inc

```diff
@@ -0,0 +1,236 @@
+<?php
+
+/**
+ * @file
+ * This file contains functions to create breadcrumbs using Solr.
+ */
+
+/**
+ * Implements callback_islandora_breadcrumbs_backends.
+ */
+function islandora_solr_islandora_breadcrumbs_backends_callable($object) {
+  $breadcrumbs = islandora_solr_get_breadcrumbs_recursive($object->id);
+  array_pop($breadcrumbs);
+  return $breadcrumbs;
+}
+
+/**
+ * Loads a map of collection hierarchy.
+ *
+ * @return array
+ *   The array has current pid as key pointing to an associative array of
+ *   'parent' => pid and 'label' => parent_label.
+ */
+function islandora_solr_load_crumbs() {
+  if (user_is_anonymous()) {
+    $cache_roles = 'anonymous';
+  }
+  else {
+    global $user;
+    $roles = $user->roles;
+    // Sort so we keep ordering consistent to reuse cache.
+    asort($roles);
+    $cache_roles = implode('_', $roles);
+  }
+
+  $crumb_cache = &drupal_static(__FUNCTION__);
+
+  $breadcrumbs = array();
+
+  if (!isset($crumb_cache)) {
+    $cache = cache_get(ISLANDORA_SOLR_BREADCRUMB_CACHE_NAME, 'cache_menu');
+    if (!$cache) {
+      // Nothing in cache so initialize it with these roles.
+      $fetch = $cache_roles;
+    }
+    else {
+      $crumb_cache = $cache->data;
+    }
+  }
+
+  if (!isset($crumb_cache[$cache_roles])) {
+    // No cache for this set of roles.
+    $fetch = $cache_roles;
+  }
+  else {
+    $breadcrumbs = $crumb_cache[$cache_roles];
+  }
+
+  if (!$breadcrumbs) {
+    $fetch = $cache_roles;
+  }
+
+  if (isset($fetch)) {
+    module_load_include('inc', 'islandora_solr_search', 'includes/query_processor');
+    $solr_build = new IslandoraSolrQueryProcessor();
+
+    $content_model = variable_get('islandora_solr_content_model_field', 'RELS_EXT_hasModel_uri_ms');
+    $collection_field = variable_get('islandora_solr_member_of_collection_field', 'RELS_EXT_isMemberOfCollection_uri_ms');
```

Comment:
Also...and I apologize for not noticing this sooner. This whole `islandora_solr_load_crumbs()` actually creates a map of Collection -> parent Collection which is why I used that predicate. So (and this might be where I went wrong) I assume that once you have reached a Collection the parent of that collection will be another Collection.

So in the breadcrumb builder, it actually just [grabs any object relationship](https://github.com/Islandora/islandora_solr_search/pull/292/files/063860d2b9d33cc1a566ff890969c9e041da5b90#diff-b777cefeb51b908c370b342ac9947778R221). I can replace that with a specific check. 

But anyways, so if you are on a Book Page it uses that function to traverse to the Book and again from Book to the Collection. Once on a Collection it uses the cached map to traverse Collections up to islandora:root. 

I did that thinking that you are usually only 1 or 2 relationships from a Collection, but you could be several Collections deep. So caching the Collection map saved time.